File: 03-Socket-TCP/esempi/es01-maiusc/TcpServer.java
/**
 * Esempio di un semplice server TCP che riceve un messaggio da un client,
 * lo elabora e restituisce una risposta con il messaggio in maiuscolo.
 *
 * javac TcpServer.java && java TcpServer
 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;

public class TcpServer {
    public static void main(String[] args) {
        try (
            ServerSocket serverSocket = new ServerSocket(8765);
            Socket clientSocket = serverSocket.accept(); // Attende la connessione di un client
            BufferedReader in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
            PrintWriter out = new PrintWriter(clientSocket.getOutputStream(), true)
        ) {
            System.out.println("Server in ascolto sulla porta 8765...");
            System.out.println("Connessione stabilita con il client!");
            String message = in.readLine(); // Legge il messaggio dal client
            System.out.println("Messaggio ricevuto: " + message);
            out.println("Risposta dal server: " + message.toUpperCase());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: 03-Socket-TCP/esempi/es01-maiusc/TcpServer.js
/**
 * Esempio di un semplice server TCP che riceve un messaggio da un client,
 * lo elabora e restituisce una risposta con il messaggio in maiuscolo.
 *
 * Eseguo il server in background: 
 * node TcpServer.js &
 */
const net = require('net');

const PORT = 8765;

// Crea il server TCP
const server = net.createServer((socket) => {
    console.log('Connessione stabilita con il client!');
    console.log(`Client: ${socket.remoteAddress}:${socket.remotePort}`);

    // Gestione dati ricevuti dal client
    socket.on('data', (data) => {
        const message = data.toString().trim();
        console.log('Messaggio ricevuto: ' + message);
        
        // Invia la risposta al client (messaggio in maiuscolo)
        const response = 'Risposta dal server: ' + message.toUpperCase() + '\n';
        socket.write(response);
        
        // Chiude la connessione dopo aver inviato la risposta
        socket.end();
    });

    // Gestione chiusura connessione
    socket.on('end', () => {
        console.log('Client disconnesso');
        server.close(() => {
            console.log('Server terminato.');
        });
    });

    // Gestione errori
    socket.on('error', (err) => {
        console.error('Errore socket:', err.message);
    });
});

// Gestione errori del server
server.on('error', (err) => {
    console.error('Errore server:', err.message);
});

// Avvia il server
server.listen(PORT, () => {
    console.log(`Server in ascolto sulla porta ${PORT}...`);
});

/*
Notare che il server rimane in esecuzione in background 
    (senza il codice di riga 36-40)
    server.close(() => {
        console.log('Server terminato.');
    });
e può gestire più connessioni sequenziali dai client.
per terminare il server, usare il comando:
fg # per portare il processo in foreground
kill %1
oppure
premere Ctrl+C nel terminale.

*/